File: java/328.odd-even-linked-list.java
/*
 * @lc app=leetcode id=328 lang=java
 *
 * [328] Odd Even Linked List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode oddHead = head, oddPtr = oddHead;
        ListNode evenHead = head.next, evenPtr = evenHead;
        head = evenHead.next;
        boolean isOdd = true;
        while (head != null) {
            if (isOdd) {
                oddPtr.next = head;
                oddPtr = oddPtr.next;
            } else {
                evenPtr.next = head;
                evenPtr = evenPtr.next;
            }
            head = head.next;
            isOdd = !isOdd;
        }
        oddPtr.next = evenHead;
        evenPtr.next = null;
        return oddHead;
    }
}

File: go/328.odd-even-linked-list.go
/*
 * @lc app=leetcode id=328 lang=golang
 *
 * [328] Odd Even Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func oddEvenList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
		return head
	}
	oddahead, evenahead := new(ListNode), new(ListNode)
	cur, oddhead, evenhead := head, oddahead, evenahead
	for i := 1; cur != nil; i, cur = i+1, cur.Next {
		if i % 2 == 1 {
			oddhead.Next = cur
			oddhead = cur
		} else {
			evenhead.Next = cur
			evenhead = cur
		}
	}
	evenhead.Next = nil
	oddhead.Next = evenahead.Next
	return oddahead.Next
}
// @lc code=end

